### ISSUE TITLE
Taint check failure

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4008]
Migrated from [rt.perl.org#7015](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7015) (status was 'resolved')

Searchable as RT7015$

**Comment by p5pRT** [comment]
### From me-01@ton.iguana.be



Consider this program&#8203;:
\-\-\-
\#\! /usr/bin/perl \-lwT
use Fcntl;
sub is\_tainted \{
&nbsp;    eval \{
&nbsp;        my $dummy = join\(''\,@&#8203;\_\)\, kill 0;
&nbsp;        1;
&nbsp;    \} ? " is not tainted" : " is tainted";
\}

my $test = shift\(@&#8203;ARGV\);
my $name = "blub\.$test";
my $untaint = 4;
print STDERR "variable $untaint"\, is\_tainted\($untaint\);
print STDERR "variable $name"\, is\_tainted\($name\);
if \($test == 0\) \{
&nbsp;    open\(my $fh\, ">"\, $name\) || die $\!;
\} else \{
&nbsp;    sysopen\(my $fh\, $name\, O\_RDWR | O\_CREAT\) || die $\!;
\}
\-\-\-
it creates a file\, once with sysopen \(if you give input 1\)\, 
once with open \(if you give input 0\)\. Both should fail\. However\, 
the sysopen method succeeds\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.7.1:

Configured by ton at Sat May 12 18:42:24 CEST 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.3, archname=i686-linux
    uname='linux quasar 2.4.3 #19 sun apr 1 10:46:29 cest 2001 i686 unknown '
    config_args='-e'
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING',
    optimize='-g',
    cppflags='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING'
    ccversion='', gccversion='egcs-2.91.66 19990314 (egcs-1.1.2 release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib -L/opt/gnu/lib'
    libpth=/usr/local/lib /opt/gnu/lib /lib /usr/lib
    libs=-lnsl -lndbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.1.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.7.1/i686-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib -L/opt/gnu/lib'

Locally applied patches:
    


@INC for perl v5.7.1:
    /usr/lib/perl5/5.7.1/i686-linux
    /usr/lib/perl5/5.7.1
    /usr/lib/perl5/site_perl/5.7.1/i686-linux
    /usr/lib/perl5/site_perl/5.7.1
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.7.1:
    HOME=/home/ton
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ton/bin.Linux:/home/ton/bin:/home/ton/bin.SampleSetup:/usr/local/bin:/usr/local/sbin:/usr/local/jre/bin:/usr/local/jdk116_v5/bin:/home/oracle/product/8.1.5/bin:/usr/games/bin:/usr/X11R6/bin:/usr/share/bin:/usr/bin:/usr/sbin:/bin:/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jhi

Ouch\.  Quite embarrassing\.

Change 10172 by jhi@&#8203;alpha on 2001/05/20 20&#8203;:33&#8203;:08

&nbsp;    Fix for ID 20010519\.003&#8203;: sysopen\(\) wasn't tainting :\-\(

Affected files \.\.\.

\.\.\. //depot/perl/doio\.c\#156 edit
\.\.\. //depot/perl/t/op/taint\.t\#30 edit

Differences \.\.\.

==== //depot/perl/doio\.c\#156 \(text\) ====
Index&#8203;: perl/doio\.c
<details><summary>Inline Patch</summary>

```diff
--- perl/doio.c.~1~	Mon May 21 00:33:13 2001
+++ perl/doio.c	Mon May 21 00:33:13 2001
@@ -141,12 +141,14 @@
         /* sysopen style args, i.e. integer mode and permissions */
 	STRLEN ix = 0;
 	if (num_svs != 0) {
-	     Perl_croak(aTHX_ "panic:sysopen with multiple args");
+	     Perl_croak(aTHX_ "panic: sysopen with multiple args");
 	}
+	if (rawmode & (O_WRONLY|O_RDWR|O_APPEND|O_CREAT|O_TRUNC))
+	    TAINT_PROPER("sysopen");
 	mode[ix++] = '#'; /* Marker to openn to use numeric "sysopen" */
 
 #if defined(USE_64_BIT_RAWIO) && defined(O_LARGEFILE)
-	rawmode |= O_LARGEFILE;
+	rawmode |= O_LARGEFILE;	/* Transparently largefiley. */
 #endif
 
 #ifndef O_ACCMODE
@@ -193,7 +195,7 @@
 	num_svs = 1;
 	svp = &namesv;
         type = Nullch;
-	fp = PerlIO_openn(aTHX_ type,mode, -1, rawmode, rawperm, NULL, num_svs, svp);
+	fp = PerlIO_openn(aTHX_ type, mode, -1, rawmode, rawperm, NULL, num_svs, svp);
     }
     else {
 	/* Regular (non-sys) open */


```
</details>
==== //depot/perl/t/op/taint.t#30 (xtext) ====
Index&#8203;: perl/t/op/taint\.t
<details><summary>Inline Patch</summary>

```diff
--- perl/t/op/taint.t.~1~	Mon May 21 00:33:13 2001
+++ perl/t/op/taint.t	Mon May 21 00:33:13 2001
@@ -106,7 +106,7 @@
 close PROG;
 my $echo = "$Invoke_Perl $ECHO";
 
-print "1..155\n";
+print "1..173\n";
 
 # First, let's make sure that Perl is checking the dangerous
 # environment variables. Maybe they aren't set yet, so we'll
@@ -735,3 +735,67 @@
     close IN;
 }
 
+{
+    # bug id 20010519.003
+
+    use Fcntl;
+
+    my $evil = "foo" . $TAINT;
+
+    eval { sysopen(my $ro, $evil, O_RDONLY) };
+    test 156, $@ !~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $wo, $evil, O_WRONLY) };
+    test 157, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $rw, $evil, O_RDWR) };
+    test 158, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $ap, $evil, O_APPEND) };
+    test 159, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $cr, $evil, O_CREAT) };
+    test 160, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $tr, $evil, O_TRUNC) };
+    test 161, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $ro, "foo", O_RDONLY | $evil) };
+    test 162, $@ !~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $wo, "foo", O_WRONLY | $evil) };
+    test 163, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $rw, "foo", O_RDWR | $evil) };
+    test 164, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $ap, "foo", O_APPEND | $evil) };
+    test 165, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $cr, "foo", O_CREAT | $evil) };
+    test 166, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $tr, "foo", O_TRUNC | $evil) };
+    test 167, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $ro, "foo", O_RDONLY, $evil) };
+    test 168, $@ !~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $wo, "foo", O_WRONLY, $evil) };
+    test 169, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $rw, "foo", O_RDWR, $evil) };
+    test 170, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $ap, "foo", O_APPEND, $evil) };
+    test 171, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $cr, "foo", O_CREAT, $evil) };
+    test 172, $@ =~ /^Insecure dependency/, $@;
+
+    eval { sysopen(my $tr, "foo", O_TRUNC, $evil) };
+    test 173, $@ =~ /^Insecure dependency/, $@;
+
+    unlink("foo"); # not unlink($evil), because that would fail...
+}
+
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Great \! Thanks\.

PS&#8203;: wasn't there a discussion where the conclusion was that perlbug reports
&nbsp;    should be passed to the mailinglist ? I never saw my bugreport appear\.
&nbsp;    \(of course\, a fix as first indication that it got through is definitely
&nbsp;    nice\)
